server-utils/src/cors.rs

```diff
@@ -201,18 +246,98 @@ pub fn get_cors_header(origin: Option<&str>, host: Option<&str>, allowed: &Optio
 						}
 					})
 					.map(|_| AccessControlAllowOrigin::Value(Origin::parse(origin)))
-					.map(CorsHeader::Ok).unwrap_or(CorsHeader::Invalid)
+					.map(AllowOrigin::Ok).unwrap_or(AllowOrigin::Invalid)
 				},
 			}
 		},
 	}
 }
 
+/// Validates if the `AccessControlAllowedHeaders` in the request are allowed.
+pub fn get_cors_allow_headers(request_headers: &Headers, cors_allow_headers: &AccessControlAllowHeaders) -> AllowHeaders {
+	// Check if the header fields which were sent in the request are required
+	if let AccessControlAllowHeaders::Only(only) = cors_allow_headers {
+		let mut allowed_headers = only.clone();
+		allowed_headers.extend(get_always_allowed_headers().iter().cloned());
+
+		for header in request_headers.iter() {
+			let name = &Ascii::new(header.name().to_owned());
+			if !allowed_headers.contains(name) {
+				return AllowHeaders::Invalid;
+			}
+		}
+	}
+
+	// Check if `AccessControlRequestHeaders` contains fields which were allowed
+	match request_headers.get::<AccessControlRequestHeaders>() {
+		None => {
+			// No fields were requested, no comparison necessary
+			match cors_allow_headers {
+				AccessControlAllowHeaders::Any => AllowHeaders::NotRequired,
+
+				// The fields which can be used are constrainted, but since it
+				// was asked for none we don't return any.
+				AccessControlAllowHeaders::Only(_) => {
+					let empty = header::AccessControlAllowHeaders(vec![]);
+					AllowHeaders::Ok(empty)
+				},
+			}
+		},
+		Some(requested) => {
+			// "requested" contains the fields for which it is inquired to know
+			// if they can be used.
+
+			let echo = AllowHeaders::Ok(
+				header::AccessControlAllowHeaders(requested.deref().clone())
+			);
+
+			match cors_allow_headers {
+				AccessControlAllowHeaders::Any => {
+					// Any field is allowed. Our response are the fields about which
+					// the request inquired.
+					echo
+				},
+
+				AccessControlAllowHeaders::Only(only) => {
+					let mut allowed_headers = only.clone();
```

Comment:
Why do we need to clone it with every request? Maybe just do two checks instead of extending and doing one?

```rust
let are_all_allowed = requested.iter().all(|header| only.contains(header) || always_allowed.contains(header));
```

Reply:
Ah yeah, you're right, the cloning is not necessary. I replaced it with your suggestion.